vshard/router/init.lua

```diff
@@ -978,15 +980,19 @@ local function replicasets_map_reduce(replicasets, rid, func, args,
     -- Map stage: send.
     --
     local func_args = {'storage_map', rid, func, args}
+    local grouped_args_added = false
     for id, rs in pairs(replicasets) do
         if grouped_args ~= nil then
             -- It's cheaper to push and then pop, rather then deepcopy
             -- arguments table for every call.
-            table.insert(args, grouped_args[id])
+            local rs_grouped_args = grouped_args[id]
+            table.insert(args, rs_grouped_args)
+            grouped_args_added = rs_grouped_args ~= nil
         end
         local res, err = rs:callrw('vshard.storage._call', func_args, opts_map)
-        if grouped_args ~= nil then
+        if grouped_args_added then
             table.remove(args)
+            grouped_args_added = false
         end
```

Comment:
You won't need this flag I think if you would change it to something as:
```Lua
for ... do
    local rs_args = grouped_args and grouped_args[id] or nil
    if rs_args then table.insert(args, grouped_args[id]) end
    -- do the call
    if rs_args then table.remove(args) end
end
```
The code style is a bit relaxed, but you get the idea.